packages/design-system/src/components/primitives/dnd/canvas.stories.tsx

```diff
@@ -127,99 +128,147 @@ export const Canvas = () => {
-    index: number;
-    placementRect: Rect;
-  }>();
-
+  const [currentDropTarget, setCurrentDropTarget] = useState<
+    DropTarget<string> | undefined
+  >();
   const [dragItemId, setDragItemId] = useState<string>();
 
-  const dropTargetId = useRef<string>();
   const rootRef = useRef<HTMLElement | null>(null);
 
-  const setDropTarget = (id: string, element: HTMLElement) => {
-    dropTargetId.current = id;
-    placementHandlers.handleTargetChange(element);
+  const getDefaultDropTarget = () => {
+    if (rootRef.current === null) {
+      throw new Error("should not happen");
+    }
+    return { data: ROOT_ID, element: rootRef.current };
   };
 
-  const dropTargetHandlers = useDropTarget({
+  const dropHandlers = useDrop<string>({
     edgeDistanceThreshold: 10,
-    isDropTarget(element: HTMLElement) {
-      return elementToId(element) !== undefined;
+
+    isDropTarget(element) {
+      return elementToId(element) ?? false;
     },
-    onDropTargetChange(event) {
-      const id = elementToId(event.target);
+
+    swapDropTarget(dropTarget) {
       const rootElement = rootRef.current;
+      if (dropTarget === undefined || rootElement === null) {
+        return getDefaultDropTarget();
+      }
 
-      if (id === undefined || rootElement === null) {
-        return;
+      const { data: id, nearEdge } = dropTarget;
+
+      if (id === ROOT_ID) {
+        return dropTarget;
       }
 
-      if (id !== ROOT_ID) {
-        let path = findItemPath(data, id) ?? [];
-
-        // NOTE: not sure we should always do this.
-        // Might depend on whether there's a parent with acceptsChildren, or something.
-        if (event.area !== "center") {
-          path = path.slice(1);
-        }
-
-        // to make sure we are not dropping on ourself
-        const dragItemIndex = path.findIndex((item) => item.id === dragItemId);
-        if (dragItemIndex !== -1) {
-          path = path.slice(dragItemIndex + 1);
-        }
-
-        for (const item of path) {
-          if (item.acceptsChildren) {
-            const element = idToElement(rootElement, item.id);
-            if (element) {
-              setDropTarget(item.id, element);
-              return;
-            }
-          }
-        }
+      const path = findItemPath(data, id) ?? [];
+
+      if (nearEdge) {
+        path.shift();
       }
 
-      setDropTarget(ROOT_ID, rootElement);
+      // Don't allow to dpop inside drag item or any of its children
+      const dragItemIndex = path.findIndex((item) => item.id === dragItemId);
+      if (dragItemIndex !== -1) {
+        path.splice(0, dragItemIndex + 1);
+      }
+
+      const newItem = path.find((item) => item.acceptsChildren);
+
+      if (newItem === undefined) {
+        return getDefaultDropTarget();
+      }
+
+      const element = idToElement(rootElement, newItem.id);
+
+      if (element === undefined) {
+        return getDefaultDropTarget();
+      }
+
+      return { data: newItem.id, element };
```

Comment:
I wonder if we could reuse this somehow with the actual logic,  seems like this stuff may break between storybook and designer

Reply:
Do you mean reuse the logic inside swapDropTarget? Even between the 3 places where we use the hook it's kinda unique in each place. I'd kept it in the app code for now rather than trying to move into the dnd hooks.